hadoop-tools/hadoop-azure/src/test/java/org/apache/hadoop/fs/azurebfs/ITestAzureBlobFileSystemCreate.java

```diff
@@ -927,6 +929,64 @@ public void testCreateNonRecursive2() throws Exception {
     assertIsFile(fs, testFile);
   }
 
+  @Test
+  public void testCreateNonRecursiveForAtomicDirectoryFile() throws Exception {
+    AzureBlobFileSystem fileSystem = getFileSystem();
+    Assume.assumeTrue(fileSystem.getAbfsStore().getAbfsConfiguration().getPrefixMode() == PrefixMode.BLOB);
+    fileSystem.setWorkingDirectory(new Path("/"));
+    fileSystem.mkdirs(new Path("/hbase/dir"));
+    fileSystem.createFile(new Path("/hbase/dir/file"))
+        .overwrite(false)
+        .replication((short) 1)
+        .bufferSize(1024)
+        .blockSize(1024)
+        .build();
+    Assert.assertTrue(fileSystem.exists(new Path("/hbase/dir/file")));
+  }
+
+  @Test
+  public void testActiveCreateNonRecursiveDenyParallelReadOnAtomicDir() throws Exception {
+    AzureBlobFileSystem fileSystem = (AzureBlobFileSystem) FileSystem.newInstance(getRawConfiguration());
+    AbfsClient client = Mockito.spy(fileSystem.getAbfsClient());
+    fileSystem.getAbfsStore().setClient(client);
+    Assume.assumeTrue(fileSystem.getAbfsStore().getAbfsConfiguration().getPrefixMode() == PrefixMode.BLOB);
+    fileSystem.setWorkingDirectory(new Path("/"));
+    fileSystem.mkdirs(new Path("/hbase/dir"));
+    fileSystem.create(new Path("/hbase/dir/file"));
+    AtomicBoolean createCalled = new AtomicBoolean(false);
+    AtomicBoolean parallelRenameDone = new AtomicBoolean(false);
+    AtomicBoolean exceptionCaught = new AtomicBoolean(false);
+
+    Mockito.doAnswer(answer -> {
+      AbfsRestOperation op = (AbfsRestOperation) answer.callRealMethod();
+      createCalled.set(true);
+      while(!parallelRenameDone.get());
+      return op;
+    }).when(client).createPathBlob(Mockito.anyString(), Mockito.anyBoolean(),
+        Mockito.anyBoolean(), Mockito.nullable(HashMap.class), Mockito.nullable(String.class), Mockito.nullable(TracingContext.class));
+
+    new Thread(() -> {
```

Comment:
Use of ExecutorService or CompletableFuture

Reply:
It will not make any difference here. Generally in test, completableFutures are used, so that there could be a way to ensure parallel thread complete before test exists.
Here, in the parallel thread, we set parallelRenameDone for which main thread (test thread) is waiting in :
```
    Mockito.doAnswer(answer -> {
      AbfsRestOperation op = (AbfsRestOperation) answer.callRealMethod();
      createCalled.set(true);
      while(!parallelRenameDone.get());
      return op;
    }).when(client).createPathBlob(Mockito.anyString(), Mockito.anyBoolean(),
        Mockito.anyBoolean(), Mockito.nullable(HashMap.class), Mockito.nullable(String.class), Mockito.nullable(TracingContext.class));

```